<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Изучение-данных-из-файла" data-toc-modified-id="Изучение-данных-из-файла-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Изучение данных из файла</a></span></li><li><span><a href="#Проверка-на-наличие-явных-дубликатов" data-toc-modified-id="Проверка-на-наличие-явных-дубликатов-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Проверка на наличие явных дубликатов</a></span></li><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Преобразование-данных" data-toc-modified-id="Преобразование-данных-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Преобразование данных</a></span></li><li><span><a href="#Первичное-обучение-моделей" data-toc-modified-id="Первичное-обучение-моделей-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Первичное обучение моделей</a></span></li><li><span><a href="#Проверка-адекватности" data-toc-modified-id="Проверка-адекватности-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Проверка адекватности</a></span></li><li><span><a href="#Матрица-ошибок,-полнота,-точность,-F1-мера" data-toc-modified-id="Матрица-ошибок,-полнота,-точность,-F1-мера-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Матрица ошибок, полнота, точность, F1-мера</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.4.2"><span class="toc-item-num">2.4.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.4.3"><span class="toc-item-num">2.4.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.4.4"><span class="toc-item-num">2.4.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span></li><li><span><a href="#Обучение-моделей-на-сбаланированной-выборке" data-toc-modified-id="Обучение-моделей-на-сбаланированной-выборке-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Обучение моделей на сбаланированной выборке</a></span></li><li><span><a href="#Улучшение-модели-случайного-леса" data-toc-modified-id="Улучшение-модели-случайного-леса-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Улучшение модели случайного леса</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Проверка-на-адекватность-тестовой-модели" data-toc-modified-id="Проверка-на-адекватность-тестовой-модели-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Проверка на адекватность тестовой модели</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Подготовка данных

In [1]:
import pandas as pd
import math
import numpy as np
from scipy import stats as st
from IPython.display import display
import matplotlib.pyplot as plt
import matplotlib.patches as patches
from pymystem3 import Mystem
import seaborn as sns
import random

from sklearn.utils import shuffle
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.dummy import DummyClassifier
from sklearn.preprocessing import StandardScaler


from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve



import warnings
warnings.filterwarnings('ignore')

pd.options.display.max_columns = None
pd.options.display.max_rows = None
pd.set_option('display.float_format', '{:.2f}'.format)
pd.options.mode.chained_assignment = None 

### Изучение данных из файла

In [2]:
data = pd.read_csv('/datasets/Churn.csv')

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [4]:
data.sample(7,random_state = 123)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
2656       2657    15713267    Zimmer          779     Spain  Female   34   
445         446    15803406      Ross          748    France  Female   26   
9505       9506    15571281     Ts'ao          651    France    Male   28   
332         333    15702669  Faulkner          663   Germany    Male   44   
4168       4169    15690796  Chambers          516    France    Male   37   
2364       2365    15796343  Bazhenov          707    France  Female   31   
6097       6098    15615050    Savage          575   Germany    Male   47   

      Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
2656    5.00      0.00              2          0               1   
445     1.00  77780.29              1          0               1   
9505   10.00  79562.98              1          1               1   
332     2.00 117028.60              2          0               1   
4168    8.00      0.00              1          1               0   
2364    2.00  82787.93              2          0               0   
6097    9.00 107915.94              2          1               1   

      EstimatedSalary  Exited  
2656        111676.63       0  
445         183049.41       0  
9505         74687.37       0  
332         144680.18       0  
4168        101834.58       0  
2364         91423.69       0  
6097         63452.18       1

### Проверка на наличие явных дубликатов

In [5]:
data.duplicated().sum()

0

### Обработка пропусков

In [6]:
def check_of_passes(df):
    """
    Осмотр датасета на структуру пропусков
    Показывает какой процент данных отсутствует
    
    """
    a = df.isna().sum().sort_values(ascending=False)
    return  (((a/len(df))*100).round(0).astype(int)).astype(str) + ' %'

In [7]:
check_of_passes(data)

Tenure             9 %
RowNumber          0 %
CustomerId         0 %
Surname            0 %
CreditScore        0 %
Geography          0 %
Gender             0 %
Age                0 %
Balance            0 %
NumOfProducts      0 %
HasCrCard          0 %
IsActiveMember     0 %
EstimatedSalary    0 %
Exited             0 %
dtype: object

В колонке `Tenure` было обнаружено **9 %** пропущеных значений. Для дальнейшие работе нам нужно избавится от пропусков.
Заменим пропущенные значения на медианное значение столбца. Медиана для столбца `Tenure` равна **5**.

In [8]:
data = data.fillna(data['Tenure'].median())

In [9]:
check_of_passes(data)

RowNumber          0 %
CustomerId         0 %
Surname            0 %
CreditScore        0 %
Geography          0 %
Gender             0 %
Age                0 %
Tenure             0 %
Balance            0 %
NumOfProducts      0 %
HasCrCard          0 %
IsActiveMember     0 %
EstimatedSalary    0 %
Exited             0 %
dtype: object

In [10]:
#def rand_tenure(num):
    #random.seed(12345)
    #return random.randint(data['Tenure'].min(), data['Tenure'].max())

#data.loc[data['Tenure'].isnull(), 'Tenure'] = data.loc[data['Tenure'].isnull(), 'Tenure'].apply(rand_tenure)

In [11]:
data['Tenure'].value_counts(normalize=True)

5.00    0.18
1.00    0.10
2.00    0.10
8.00    0.09
3.00    0.09
7.00    0.09
4.00    0.09
9.00    0.09
6.00    0.09
10.00   0.04
0.00    0.04
Name: Tenure, dtype: float64

### Вывод

 И так данные находятся в таблице `data`:
        
Признаки:

* `RowNumber` — индекс строки в данных
* `CustomerId` — уникальный идентификатор клиента
* `Surname` — фамилия
* `CreditScore` — кредитный рейтинг
* `Geography` — страна проживания
* `Gender` — пол
* `Age` — возраст
* `Tenure` — сколько лет человек является клиентом банка
* `Balance` — баланс на счёте
* `NumOfProducts` — количество продуктов банка, используемых клиентом
* `HasCrCard` — наличие кредитной карты
* `IsActiveMember` — активность клиента
* `EstimatedSalary` — предполагаемая зарплата

Целевой признак:

* `Exited` — факт ухода клиента

## Исследование задачи

### Преобразование данных

Преобразуем категориальные признаки к количественным признакам с помощью OHE (техника прямого кодирования). Чтобы не попасть в дамми-ловушку используем функцию pd.get_dummies() с аргументом drop_first=True. Так из датасета удалим столбцы, которые не нужны нам как признаки `RowNumber`,`Surname`,`CustomerId`

In [12]:
data_ohe = data.drop(['RowNumber','Surname','CustomerId'] , axis=1)

In [13]:
data_ohe = pd.get_dummies(data_ohe, drop_first=True)

In [14]:
data_ohe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CreditScore        10000 non-null  int64  
 1   Age                10000 non-null  int64  
 2   Tenure             10000 non-null  float64
 3   Balance            10000 non-null  float64
 4   NumOfProducts      10000 non-null  int64  
 5   HasCrCard          10000 non-null  int64  
 6   IsActiveMember     10000 non-null  int64  
 7   EstimatedSalary    10000 non-null  float64
 8   Exited             10000 non-null  int64  
 9   Geography_Germany  10000 non-null  uint8  
 10  Geography_Spain    10000 non-null  uint8  
 11  Gender_Male        10000 non-null  uint8  
dtypes: float64(3), int64(6), uint8(3)
memory usage: 732.5 KB


Разделим данные на признаки и целевой признак. Так же разделим на три выборки: тренировочную, валидационную и тестовую с отношение 3:1:1. Так же масштабируем признаки, чтобы все признаки были значемы. 

In [15]:
target = data_ohe['Exited']
features = data_ohe.drop('Exited', axis=1)
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.4, random_state=123) 

features_test, features_valid, target_test, target_valid = train_test_split(features_test,target_test,
                                                                           test_size=0.5, random_state=123)

numeric = ['CreditScore', 'Tenure', 'Balance', 'Age', 'NumOfProducts','EstimatedSalary']

scaler = StandardScaler()
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])

print(features_train.shape) 
print(features_valid.shape) 
print(features_test.shape) 


(6000, 11)
(2000, 11)
(2000, 11)


### Первичное обучение моделей

Проверим наше данные. Попробуем обучить данные с помощью решающего деревья, логистической регрессии и случайного леса.

In [16]:
model_DTC = DecisionTreeClassifier(random_state=12345)
DTC_score = model_DTC.fit(features_train, target_train).score(features_valid, target_valid)

model_RFC = RandomForestClassifier(random_state=12345, n_estimators = 40, max_depth =13)
RFC_score = model_RFC.fit(features_train, target_train).score(features_valid, target_valid)

model_LGR = LogisticRegression(random_state=12345, solver = 'liblinear')
LGR_score = model_LGR.fit(features_train, target_train).score(features_valid, target_valid)
print("Точность:" "дерево решений", DTC_score, "случайный лес ", RFC_score, "логистическая регрессия", LGR_score)

Точность:дерево решений 0.8065 случайный лес  0.8665 логистическая регрессия 0.8215


Проверим баланс классов.

In [17]:
target_valid.value_counts(normalize = 1)

0   0.80
1   0.20
Name: Exited, dtype: float64

У обучающией выбороки наблюдается заметный дисбаланс класов. Ответы **0 ≈80%** , **1 ≈ 20%**.
Можно ожидать, что результаты предсказаний моделей будут сильно склоняться к одному варианту ответов.

### Проверка адекватности

In [18]:
model_DTC = DecisionTreeClassifier(random_state=123)
model_DTC.fit(features_train, target_train)
DTC_share = pd.Series(model_DTC.predict(features_valid)).value_counts(normalize = 1)


model_RFC = RandomForestClassifier(random_state=12345, n_estimators = 40, max_depth =13)
model_RFC.fit(features_train, target_train)
RFC_share = pd.Series(model_RFC.predict(features_valid)).value_counts(normalize = 1)

model_LgR = LogisticRegression(solver = 'liblinear')
model_LgR.fit(features_train, target_train)
LgR_share = pd.Series(model_LgR.predict(features_valid)).value_counts(normalize = 1)
print("Доли ответов:" "дерево решений", DTC_share, "случайный лес ", RFC_share, "логистческая регрессия", LgR_share , end='')

Доли ответов:дерево решений 0   0.80
1   0.20
dtype: float64 случайный лес  0   0.88
1   0.12
dtype: float64 логистческая регрессия 0   0.92
1   0.08
dtype: float64

Воспльзуемся моделью `DummyClassifier` со стратегией `most_frequent`чтобы сравнить качество модели.

In [19]:
dummy_clf = DummyClassifier(strategy="most_frequent")
dummy_clf.fit(features_train, target_train).score(features_valid, target_valid)

0.798

По метрики `accuracy` все отлично. Все наши модели по точности больше модели `DummyClassifier`. Посмотрим для каждой модели матрицу ошибок.

### Матрица ошибок, полнота, точность, F1-мера

#### Дерево решений 

In [21]:
model_DTC = DecisionTreeClassifier(random_state=123)
model_DTC.fit(features_train, target_train)
DTC_prediction = model_DTC.predict(features_valid)
print(DTC_prediction)
probabilities_valid_DTC = model_DTC.predict_proba(features_valid)
probabilities_one_valid_DTC = probabilities_valid_DTC[:, 1]
print(probabilities_valid_DTC)
confusion_matrix(target_valid, DTC_prediction)

[1 0 0 ... 1 0 1]
[[0. 1.]
 [1. 0.]
 [1. 0.]
 ...
 [0. 1.]
 [1. 0.]
 [0. 1.]]


array([[1413,  183],
       [ 191,  213]])

Дерево решений склонно выдавать позитивные предсказания (TP), очень высокое количество ложных позитивных предсказания (FP). Рассмотрим полноту,точноть и F1-меру.

In [21]:
print("Полнота" , recall_score(target_valid, DTC_prediction))
print("Точность", precision_score(target_valid, DTC_prediction))
print("F1-мера", f1_score(target_valid, DTC_prediction))
print("AUC-ROC", roc_auc_score(target_valid, probabilities_one_valid_DTC))

Полнота 0.5272277227722773
Точность 0.5378787878787878
F1-мера 0.5325000000000001
AUC-ROC 0.7062830343184695


Полнота и тончность почти одинаковы. F1-мера получилось меньше нашего минимума **0.59**.

#### Случайный лес

In [22]:
model_RFC = RandomForestClassifier(random_state=12345, n_estimators = 40, max_depth =13)
model_RFC.fit(features_train, target_train)

probabilities_valid_RFC = model_RFC.predict_proba(features_valid)
probabilities_one_valid_RFC = probabilities_valid_RFC[:, 1]

RFC_prediction = model_RFC.predict(features_valid)

confusion_matrix(target_valid, RFC_prediction)

array([[1543,   53],
       [ 214,  190]])

Модель случайного леса сильно склоняется к ложно позитивным предсказанием. Рассмотрим полноту,точноть и F1-меру.

In [23]:
print("Полнота" , recall_score(target_valid, RFC_prediction))
print("Точность", precision_score(target_valid, RFC_prediction))
print("F1-мера", f1_score(target_valid, RFC_prediction))
print("AUC-ROC", roc_auc_score(target_valid, probabilities_one_valid_RFC))

Полнота 0.47029702970297027
Точность 0.7818930041152263
F1-мера 0.5873261205564142
AUC-ROC 0.8525994751730811


Модель случайный лес показала прирост в точности и F1-мера стала больше. Все еще меньше нашего целевого значения.

#### Логистическая регрессия

In [24]:
model_LgR = LogisticRegression(solver = 'liblinear')
model_LgR.fit(features_train, target_train)

probabilities_valid_LgR = model_LgR.predict_proba(features_valid)
probabilities_one_valid_LgR = probabilities_valid_RFC[:, 1]

LgR_prediction = model_LgR.predict(features_valid)

confusion_matrix(target_valid, LgR_prediction)

array([[1536,   60],
       [ 297,  107]])

Модель случайного леса сильно склоняется к ложно позитивным предсказанием. Рассмотрим полноту,точноть и F1-меру.

In [25]:
print("Полнота" , recall_score(target_valid, LgR_prediction))
print("Точность", precision_score(target_valid, LgR_prediction))
print("F1-мера", f1_score(target_valid, LgR_prediction))
print("AUC-ROC", roc_auc_score(target_valid, probabilities_one_valid_LgR))

Полнота 0.26485148514851486
Точность 0.6407185628742516
F1-мера 0.3747810858143608
AUC-ROC 0.8525994751730811


#### Вывод

При иследование задачи выяснилось что в данных наблюдается дисбаланс классов. Кол-во отрицательных классов гораздо больше, чем положительных. 

## Борьба с дисбалансом

### Борьба с дисбалансом

Для борьбы с дисбалансом воспользуемся техникой upsampling. То есть увеличим данные положительного класса.

In [26]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

In [27]:
target_upsampled.value_counts(normalize = 1)

1   0.51
0   0.49
Name: Exited, dtype: float64

In [28]:
features_upsampled.shape

(9717, 11)

In [29]:
target_upsampled.shape

(9717,)

В этой нелегкой  борьбе мы победили. Классов стало поровну.

### Обучение моделей на сбаланированной выборке

Точность  моделей на выборке с дисбалансом:

In [30]:
model_DTC = DecisionTreeClassifier(random_state=12345)
DTC_score = model_DTC.fit(features_train, target_train).score(features_valid, target_valid)

model_RFC = RandomForestClassifier(random_state=12345, n_estimators = 40, max_depth =13)
RFC_score = model_RFC.fit(features_train, target_train).score(features_valid, target_valid)

model_LGR = LogisticRegression(random_state=12345, solver = 'liblinear')
LGR_score = model_LGR.fit(features_train, target_train).score(features_valid, target_valid)
print("Точность:" "дерево решений", DTC_score, "случайный лес ", RFC_score, "логистическая регрессия", LGR_score)

Точность:дерево решений 0.8065 случайный лес  0.8665 логистическая регрессия 0.8215


Точность  моделей на сбалансированной выборке:

In [31]:
model_DTC = DecisionTreeClassifier(random_state=12345)
DTC_score = model_DTC.fit(features_upsampled, target_upsampled).score(features_valid, target_valid)

model_RFC = RandomForestClassifier(random_state=12345, n_estimators = 40, max_depth =13)
RFC_score = model_RFC.fit(features_upsampled, target_upsampled).score(features_valid, target_valid)

model_LGR = LogisticRegression(random_state=12345, solver = 'liblinear')
LGR_score = model_LGR.fit(features_upsampled, target_upsampled).score(features_valid, target_valid)
print("Точность:" "дерево решений", DTC_score, "случайный лес ", RFC_score, "логистическая регрессия", LGR_score)

Точность:дерево решений 0.8055 случайный лес  0.851 логистическая регрессия 0.714


Точность моделей изменилась незначительно. Проверим другие метрики.

Для модели решающего дерева.

In [32]:
model_DTC = DecisionTreeClassifier(random_state=123)
model_DTC.fit(features_upsampled, target_upsampled)
DTC_prediction = model_DTC.predict(features_valid)

probabilities_valid_DTC = model_DTC.predict_proba(features_valid)
probabilities_one_valid_DTC = probabilities_valid_DTC[:, 1]

confusion_matrix(target_valid, DTC_prediction)

array([[1411,  185],
       [ 202,  202]])

In [33]:
print("Полнота" , recall_score(target_valid, DTC_prediction))
print("Точность", precision_score(target_valid, DTC_prediction))
print("F1-мера", f1_score(target_valid, DTC_prediction))
print("AUC-ROC", roc_auc_score(target_valid, probabilities_one_valid_DTC))

Полнота 0.5
Точность 0.5219638242894057
F1-мера 0.5107458912768648
AUC-ROC 0.6920426065162907


Для модели случайного леса.

In [34]:
model_RFC = RandomForestClassifier(random_state=12345, n_estimators = 40, max_depth =13)
model_RFC.fit(features_upsampled, target_upsampled)

probabilities_valid_RFC = model_RFC.predict_proba(features_valid)
probabilities_one_valid_RFC = probabilities_valid_RFC[:, 1]

RFC_prediction = model_RFC.predict(features_valid)

confusion_matrix(target_valid, RFC_prediction)

array([[1444,  152],
       [ 146,  258]])

In [35]:
print("Полнота" , recall_score(target_valid, RFC_prediction))
print("Точность", precision_score(target_valid, RFC_prediction))
print("F1-мера", f1_score(target_valid, RFC_prediction))
print("AUC-ROC", roc_auc_score(target_valid, probabilities_one_valid_RFC))

Полнота 0.6386138613861386
Точность 0.6292682926829268
F1-мера 0.6339066339066339
AUC-ROC 0.846845610312911


Для модели логистической регрессии.

In [36]:
model_LgR = LogisticRegression(solver = 'liblinear')
model_LgR.fit(features_upsampled, target_upsampled)

probabilities_valid_LgR = model_LgR.predict_proba(features_valid)
probabilities_one_valid_LgR = probabilities_valid_RFC[:, 1]

LgR_prediction = model_LgR.predict(features_valid)

confusion_matrix(target_valid, LgR_prediction)



array([[1134,  462],
       [ 110,  294]])

In [37]:
print("Полнота" , recall_score(target_valid, LgR_prediction))
print("Точность", precision_score(target_valid, LgR_prediction))
print("F1-мера", f1_score(target_valid, LgR_prediction))
print("AUC-ROC", roc_auc_score(target_valid, probabilities_one_valid_LgR))

Полнота 0.7277227722772277
Точность 0.3888888888888889
F1-мера 0.506896551724138
AUC-ROC 0.846845610312911


Показатели всех моделей улучшились.
Лучшие результаты показывает модель случайный лес. На валидационной выборке показывает результаты **F1 меры = 0.634**, что выше целевого значения.
Попробуем улучшить модель путем изменения параметров.

Теперь поработаем только с одной моделью случайного леса и подберем гиперпараметры, чтобы улучшить значения метрик.

### Улучшение модели случайного леса

In [38]:
%%time
best_model = None
best_result_recall = 0
best_result_precision = 0
best_result_f1 = 0
best_result_auc = 0 
best_est = 0
best_depth = 0
for est in range(10, 101, 10):
    for depth in range (1, 15, 2):
        model = RandomForestClassifier(random_state=12345, max_depth= depth,n_estimators=est)#
        model.fit(features_upsampled,target_upsampled)

        probabilities_valid = model.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        result_auc = roc_auc_score(target_valid, probabilities_one_valid)
        
        for threshold in np.arange(0, 0.5, 0.01):
            predicted_valid = probabilities_one_valid > threshold

            result_recall = recall_score(target_valid, predicted_valid)
            result_precision = precision_score(target_valid, predicted_valid)
            result_f1 = f1_score(target_valid, predicted_valid)
            #result_auc = roc_auc_score(target_valid, predicted_valid)

            if result_f1 > best_result_f1:
                best_result_f1 = result_f1
                best_est = est
                best_depth = depth
                best_result_recall = result_recall
                best_result_precision = result_precision
                best_result_auc = result_auc

print('est = {:.0f}, depth = {:.0f}'.format(best_est,best_depth))
print(' Recall = {:.3f}, Precision = {:.3f}, AUC = {:.3f}, F1 = {:.3f} - Threshold = {:.2f}'.format(
                        best_result_recall,best_result_precision,best_result_auc,best_result_f1, threshold ))
                    

est = 40, depth = 13
 Recall = 0.646, Precision = 0.620, AUC = 0.847, F1 = 0.633 - Threshold = 0.49
CPU times: user 40.7 s, sys: 110 ms, total: 40.8 s
Wall time: 40.9 s


### Вывод

При обнаржунеие дисбаланса в данных был применем метод **upsampling** тоесть мы увеличили выборку с помощью увелечение кол-во положительного класса. При обучение моделей на сбалансированных данных наилушчий результат показала модель случайного леса.
При подборе параметров для модели случайного леса получили значение F1-мера **0.633**, что является большего нашего целового значения. При параметрах кол-во дереьвьев **40**, глубина дерева **13** и порогом **0,49**.

## Тестирование модели

Проверим нашу модель на тестовой выборке. Для начала объединим наши две выборки между собой train+valid. Но выборка valid имеет дисбаланс, так же как и для выборки train преобразуем выборку спомощью метода upsampling.

In [39]:
target_valid.value_counts(normalize=True)

0   0.80
1   0.20
Name: Exited, dtype: float64

In [40]:
features_upsampled_valid, target_upsampled_valid = upsample(features_valid, target_valid, 4)

In [41]:
target_upsampled_valid.value_counts(normalize=True)

1   0.50
0   0.50
Name: Exited, dtype: float64

In [42]:
target_upsampled_valid.shape

(3212,)

In [43]:
target_upsampled.shape

(9717,)

In [44]:
features_upsampled_general = pd.concat([features_upsampled_valid, features_upsampled])

In [45]:
target_upsampled_general = pd.concat([target_upsampled_valid, target_upsampled])

In [46]:
features_upsampled_general.shape

(12929, 11)

In [47]:
target_upsampled_general.shape

(12929,)

In [52]:

best_model = None
best_result_recall = 0
best_result_precision = 0
best_result_f1 = 0
best_result_auc = 0 
best_est = 0
best_depth = 0


model = RandomForestClassifier(random_state=12345, max_depth= 13,n_estimators=40)
model.fit(features_upsampled_general,target_upsampled_general)

probabilities_test = model.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]

result_auc_test = roc_auc_score(target_test, probabilities_one_test)

predicted_test = probabilities_one_test > 0.49

result_recall_test = recall_score(target_test, predicted_test)
result_precision_test = precision_score(target_test, predicted_test)
result_f1_test = f1_score(target_test, predicted_test)
# result_auc_test = roc_auc_score(target_test, predicted_test)

result = model.score(features_test, target_test)
           
print(' Recall = {:.3f}, Precision = {:.3f}, AUC = {:.3f}, F1 = {:.3f}'.format(result_recall_test,
                                                                result_precision_test, result_auc_test, result_f1_test))
print('Accuracy =',result)

 Recall = 0.650, Precision = 0.591, AUC = 0.867, F1 = 0.619
Accuracy = 0.847


Модель показала отличные результат метрика **F1** превысила целевое значение **0.619** > **0.59**.

###  Проверка на адекватность тестовой модели

In [49]:
dummy_clf_test = DummyClassifier(strategy="most_frequent")
dummy_clf_test.fit(features_upsampled_general,target_upsampled_general).score(features_test, target_test)

0.197

In [56]:
#КОД РЕВЬЮЕРА
dummy_clf_test = DummyClassifier(strategy="most_frequent")
dummy_clf_test.fit(features_upsampled_general,target_upsampled_general)

predicted_dummy = dummy_clf_test.predict(features_test)
f1_score(target_test, predicted_dummy)

0.329156223893066

Проверка на адекватность пройдена. Accuracy модели случайной лес больше точности модели DummyClassifier **0.847** > **0.197**.

## Вывод

В нашем распоряжение были данные из «Бета-Банка», нужно было спрогноизировать отток клиентов на основе исторических данных. При обработке данных было обнаружено пропуск данных в столбце `Tenuare`, пропуски были заполнены случайными образ (либо `0` либо `1`). Также данные были преобразованы с помощью прямого кодирования `OHE` с аргументом `drop_first=True`, чтобы не попасть в дамми-ловушку. Данные были разделены на три выборки: тренировочную, валидационную и тестовую с отношение 3:1:1. Так же воспользовались  масштабириованием признаки, чтобы все признаки были значемы.

Для прогнозирования оттока клиентов было рассмотренно три модели: **Дерево решений**, **Случайный лес**, **Логистическая регрессия**. 

При первом изучение модели метрика **Accuracy** все три модели показали отличные результаты: **Дерево решений - 0.807**,  **Случайный лес - 0.865**,  **Логистическая регрессия -  0.8215**. Проверка на адекватность все наши три модели прошли. Для проверки на адекватность мы использовали `DummyClassifier` со стратегией `most_frequent`. Изучение других метрик таких как **матрица ошибок**, **полнота**, **точность**, **F1-мера** и **AUC-ROC** показало что не одна из моделей не показала хороших значений. И наша пороговое значение **F1-мера = 0.59** не было пройдено. Это было связано с дисбалансом данных.

Для балансирования данных был использован метод `upsampling`. Увеличи кол-во положительного класса в тренировачной выборке, так что бы положительный и отрицательный класс равен.

Обучение моделей на сбалансированных данных показал отличный результат. Метрика **Accuracy** практически не имзенилась, но вот другие метрики стали лучше. Среди моделей самый лучший резльтат метрики **F1-мера** показала модель **Случайный лес** со значением **Полнота = 0.639**, **Точность = 0.63**, **F1-мера = 0.634**, **AUC-ROC = 0.847**.

В процессе улучшения модели случайного леса были подобраны следующие гипрепарамтеры:
кол-во дереьвьев **40**, глубина дерева **13** и порогом **0,49**.

Тестирование модели на тестовой выборки показало хороший результат  **Полнота = 0.650**, **Точность = 0.591**, **AUC-ROC = 0.847**, **F1 = 0.619**. И проверка на адекватность тоже пройдена. Accuracy случайного леса **0.867** больше чем `DummyClassifier` **0.197**.

Наша цель чтобы модель показала метрику **F1 > 0.59** достигнута метрика случайного леса **F1 = 0.619**.
